### PR TITLE
feat(monobank): add monobank (UA) bank sync

### PR DESCRIPTION
**Pull request by yaroslavshparuk**
## What this adds

A **monobank** (Ukraine) provider integration — cards, jars, balances and statements — built as a per-family, bring-your-own-token provider in the same shape as the Up and Redbark integrations.

- Settings → Providers panel to connect, rotate the token, set a sync start date and disconnect.
- Account setup flow: link a monobank card/jar to an existing Sure account, create a new one from it, or skip it.
- Cards map to `Depository/checking`, jars to `Depository/savings`. A card's credit limit is subtracted out of the reported balance, so what Sure tracks is the client's own funds.
- Statement import with merchant enrichment and a category matcher mapping monobank's MCC-derived categories onto Sure categories.
- Held (unsettled) transactions import as pending and reconcile to their settled counterpart.
- English and Ukrainian locales for every new string.
- Setup guide at `docs/hosting/monobank.md`, linked from the onboarding guide.

## Why the sync design looks unusual

monobank's personal API allows **one request per minute per endpoint**, and a single statement request covers **at most 31 days**. The importer is built around those two limits rather than around throughput:

- Each sync spends a fixed budget of statement requests (`MONOBANK_MAX_STATEMENT_REQUESTS_PER_SYNC`, default 4). Accounts that don't fit are picked up by the next sync, least-recently-covered first.
- History is walked backwards across syncs (`history_synced_from`) while new activity is fetched forward from the last covered timestamp (`statement_synced_through`). Both are timestamps rather than dates: a window is requested in Unix seconds and rarely lands on midnight, so recording only the date would claim a partially covered day as complete.
- Every sync re-reads the last few days (`MONOBANK_PENDING_LOOKBACK_DAYS`, default 3) so holds stay visible in the payload instead of being pruned as stale.

`docs/hosting/monobank.md` documents all of this for operators.

## Pending handling

monobank flags a hold with `hold: true`, stored at `extra["monobank"]["pending"]`. A hold may settle under a **different** id than the hold it replaces, so the settled record reconciles onto the pending entry through `Account::ProviderImportAdapter`'s amount/date lookup. A hold that simply disappears is pruned — but only when the statement request actually covered it, and never when the entry is `protected_from_sync?` (those only lose the pending flag).

No FX metadata: the statement reports `operationAmount` but never the operation currency, so `operation_amount` is stored raw and `fx_from`/`fx_amount` are left unset.

## Incidental fix — partially addresses #3306

`Account::ProviderImportAdapter` re-hardcoded the pending-provider list in raw SQL in three places, independently of `Transaction::PENDING_PROVIDERS`. Rather than add `monobank` to a fourth copy, those three literals now derive from the constant via a single frozen `PENDING_LOOKUP_SQL`. That removes the second hardcoded list described in #3306 (item 2); the constant itself remains a hardcoded allow-list, so the issue is only partly addressed.

## Terms of use

This uses monobank's **personal** API. monobank's docs state that a service storing other people's data on its own servers must use the corporate API instead, while programs used by clients personally — where data never passes through the developer's nodes — may use the personal one. A self-hosted Sure instance is the latter, and the doc calls this out for anyone running Sure as a hosted service for others.

## Configuration

All optional, all defaulted:

| Variable | Default | Purpose |
| --- | --- | --- |
| `MONOBANK_INCLUDE_PENDING` | `1` | Import held transactions |
| `MONOBANK_MAX_STATEMENT_REQUESTS_PER_SYNC` | `4` | Statement-request budget per sync |
| `MONOBANK_PENDING_LOOKBACK_DAYS` | `3` | Overlap re-read each sync |
| `MONOBANK_INITIAL_HISTORY_DAYS` | `31` | Backfill reach for a new connection |
| `MONOBANK_DEBUG_RAW` | `0` | Raw payload logging — dev-only, PII-gated like `UP_DEBUG_RAW` |

## Testing

- `bin/rails test` — 8403 runs, 0 failures, 0 errors
- `bin/rubocop -f github` — clean
- `bundle exec erb_lint` on the new templates — clean
- `bin/brakeman --no-pager` — 0 security warnings

New coverage: provider client, importer (window splitting, budget exhaustion, pending pruning), entry processor, account processor, category matcher, controller, and the shared `ProviderImportAdapter` pending lookup.

No API endpoints were added, so no rswag/OpenAPI specs are required.

Tested against a live monobank account.

## Summary by CodeRabbit

* **New Features**
  * Added Monobank integration for Ukrainian cards and savings jars.
  * Connect with a personal API token, then select, create, or link accounts.
  * Sync balances, transactions, pending activity, and historical data.
  * Automatically categorize supported transactions using merchant category codes.
  * View connection status, sync progress, account summaries, and errors.
  * Added configurable sync options and English and Ukrainian setup documentation.

* **Bug Fixes**
  * Improved pending-transaction reconciliation and cleanup while protecting user-managed entries.
  * Improved handling of sync failures, rate limits, and account-linking conflicts.

### CODE REVIEW

File: app/views/monobank_items/select_existing_account.html.erb [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+                        data: { turbo_frame: "_top" },
+                        class: "space-y-4" do %>
+            <%= hidden_field_tag :monobank_item_id, @monobank_item.id %>
+            <%= hidden_field_tag :account_id, @account.id %>
+            <%= hidden_field_tag :return_to, @return_to %>
+
+            <div class="space-y-2">
+              <% @monobank_accounts.each do |monobank_account| %>
+                <label class="flex items-start gap-3 p-3 border border-primary rounded-lg hover:bg-surface cursor-pointer transition-colors">
+                  <%= radio_button_tag "monobank_account_id", monobank_account.id, false, class: "mt-1", required: true %>
```
[the remark this answers is not in the document]
Comment (yaroslavshparuk) [reply app/views/monobank_items/select_existing_account.html.erb:29]: The duplication is real, but it predates this PR rather than being introduced by it. The same hand-rolled row — `<label class="flex items-start gap-3 p-3 border border-primary rounded-lg hover:bg-surface cursor-pointer transition-colors">` — is already the established shape in five other places:

- `app/views/akahu_items/select_accounts.html.erb`
- `app/views/akahu_items/select_existing_account.html.erb`
- `app/views/up_items/select_accounts.html.erb`
- `app/views/up_items/select_existing_account.html.erb`
- `app/views/onchain_wallet_items/_token_review_form.html.erb`

So a `DS::*` primitive here means touching seven call sites across three unrelated providers. Doing that inside a provider integration would bury a design-system refactor in a 5k-line diff and make both halves harder to review, and matching the sibling providers is what keeps these pickers consistent today.

Happy to do the extraction as a focused follow-up PR that converts all seven at once — say the word and I will open it. If you would rather it land here, I will do it here instead; just let me know which you prefer.

File: app/models/monobank_item/importer.rb [hunk 2]
```diff
@@ -0,0 +1,523 @@
...
+      target_start = initial_history_start(monobank_account)
+      covered_from = monobank_account.history_synced_from
+      return { transactions: [] } if covered_from.blank? || covered_from <= target_start
+
+      to = covered_from
+      from = [ target_start, to - Provider::Monobank::MAX_STATEMENT_WINDOW ].max
+      return { transactions: [] } if from >= to
+
+      result = fetch_window(monobank_account, from: from, to: to, direction: :backward)
+      return { transactions: [] } unless result[:success]
```
[the remark this answers is not in the document]
Comment (yaroslavshparuk) [reply app/models/monobank_item/importer.rb]: Fixed in 5fb96d58.

`fetch_history_window` now distinguishes the two failure kinds: a throttled backward walk stays a soft skip (the forward window landed and the backfill resumes next sync, exactly as it does when the request budget runs out), while a hard failure returns the failed result so `import_transactions` counts it.

`fetch_and_store_transactions` persists the forward window *before* propagating that failure — the new activity is already in hand and dropping it would only leave the account further behind on the next sync.

Added three tests covering the outcomes: hard failure reported, forward window still persisted, rate limit still a success. Confirmed the first one fails against the previous behaviour.

File: app/models/monobank_item/importer.rb [hunk 3]
```diff
@@ -0,0 +1,523 @@
...
+          stats[:updated] += 1
+        elsif !all_existing_ids.include?(account_id.to_s)
+          monobank_account = monobank_item.monobank_accounts.build(account_id: account_id.to_s)
+          monobank_account.upsert_monobank_snapshot!(account)
+          stats[:created] += 1
+        end
+      rescue => e
+        stats[:failed] += 1
+        Rails.logger.error "MonobankItem::Importer - Failed to import account #{account_id}: #{e.message}"
+      end
```
[the remark this answers is not in the document]
Comment (yaroslavshparuk) [reply app/models/monobank_item/importer.rb:130]: Fixed in 5fb96d58. The per-account rescue now calls `capture_sync_error`, which emits a `DebugLogEntry` with `family`, `account_provider`, `provider_key` and the error class.

`capture_sync_error` gained an `extra_metadata:` parameter so the failing Monobank `account_id` is recorded even when the failure happened before the local row was persisted and there is no `MonobankAccount` to attach.

File: app/models/monobank_item/syncer.rb [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+  # Build a syncer for the given +monobank_item+.
+  def initialize(monobank_item)
+    @monobank_item = monobank_item
+  end
+
+  # Run the full sync: import, account setup detection, transaction processing,
+  # balance sync scheduling, and stats/health collection. Raises on failures.
+  def perform_sync(sync)
+    sync.update!(status_text: "Importing accounts from Monobank...") if sync.respond_to?(:status_text)
```
[the remark this answers is not in the document]
Comment (yaroslavshparuk) [reply app/models/monobank_item/syncer.rb]: Fixed in 5fb96d58. All five stages now go through `I18n.t("monobank_item.sync.status.*")`, with keys added to both `en.yml` and `uk.yml`. `needs_setup` interpolates `count` and carries the Ukrainian one/few/many/other plural forms.

Worth noting this was inherited from `UpItem::Syncer`, which still hard-codes the same five strings — `RedbarkItem`, `BrexItem` and `BinanceItem` all localize. Happy to fix Up in a follow-up if that is wanted.

File: app/models/provider/metadata.rb [hunk 5]
```diff
@@ -5,6 +5,7 @@ module Metadata
       simplefin:      { region: "US",      kinds: %w[Bank Investment], maturity: :stable, logo_text: "SF", logo_bg: "bg-blue-600" },
       lunchflow:      { region: "Global",  kinds: %w[Bank],            maturity: :stable, logo_text: "LF", logo_bg: "bg-orange-500" },
       up:             { region: "AU",      kinds: %w[Bank],            maturity: :beta,   logo_text: "UP", logo_bg: "bg-orange-600" },
+      monobank:       { region: "UA",      kinds: %w[Bank],            maturity: :alpha,  logo_text: "MB", logo_bg: "bg-black" },
```
[the remark this answers is not in the document]
Comment (yaroslavshparuk) [reply app/models/provider/metadata.rb]: Fixed in 5fb96d58 — changed to `bg-primary`.

Confirming the concern was real rather than cosmetic: the badge text is `text-inverse` (`settings/providers/_drawer_header.html.erb:8` and `settings/provider_card.html.erb:6`), which flips with the theme, so a hardcoded `bg-black` rendered dark-on-dark and made the "MB" invisible in dark mode. `bg-primary` pairs with `text-inverse` in both themes and is what `trade_republic` — also a black-branded provider — already uses in this table.

File: app/controllers/monobank_items_controller.rb [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+      return
+    end
+
+    if monobank_account.account_provider.present?
+      redirect_to accounts_path, alert: t(".monobank_account_already_linked")
+      return
+    end
+
+    AccountProvider.create!(account: account, provider: monobank_account)
+    monobank_item.sync_later
```
[the remark this answers is not in the document]
Comment (yaroslavshparuk) [reply app/controllers/monobank_items_controller.rb:218]: Good catch — fixed in 5fb96d58.

All three sites are handled:
- `link_existing_account` rescues `RecordNotUnique` around the create and returns the same `account_already_linked` flash the read guard would have given.
- `link_accounts` rescues at the action level; the transaction rolls the whole batch back, so the user goes back to the picker rather than a 500.
- `complete_account_setup` has `RecordNotUnique` added to its existing `RecordInvalid, RecordNotSaved` rescue.

File: app/controllers/settings/providers_controller.rb [hunk 7]
```diff
@@ -190,6 +190,7 @@ def reload_provider_configs(updated_fields)
     FAMILY_PANELS = [
       { key: "akahu",          title: "Akahu",           turbo_id: "akahu",          partial: "akahu_panel" },
       { key: "up",             title: "Up",              turbo_id: "up",             partial: "up_panel" },
+      { key: "monobank",       title: "monobank",        turbo_id: "monobank",       partial: "monobank_panel" },
```
[the remark this answers is not in the document]
Comment (yaroslavshparuk) [reply app/controllers/settings/providers_controller.rb:193]: Every row in `FAMILY_PANELS` passes a raw brand name — `"Akahu"`, `"Up"`, `"Lunch Flow"`, `"Redbark"`, `"SimpleFIN"`, `"Enable Banking"` and the rest. Adding a translation key for monobank alone would make it the single inconsistent entry in that table.

"monobank" is also a brand name that is not translated: the bank styles itself lowercase-latin in Ukrainian too, which is why the panel title is lowercase here.

Left as-is for consistency. If you want the table localized, it makes sense as one change converting every entry — happy to do that as a follow-up.

File: app/models/monobank_item.rb [hunk 8]
```diff
@@ -0,0 +1,222 @@
+class MonobankItem < ApplicationRecord
+  include Syncable, Provided, Unlinking, Encryptable
+
+  enum :status, { good: "good", requires_update: "requires_update" }, default: :good
+
+  if encryption_ready?
```
[the remark this answers is not in the document]
Comment (yaroslavshparuk) [reply app/models/monobank_item.rb:6]: `encryption_ready?` comes from the shared `Encryptable` concern, whose documented intent is exactly this trade-off:

> This allows encryption to be optional - if not configured, sensitive fields are stored in plaintext (useful for development or legacy deployments).

52 models use that identical `if encryption_ready?` guard, including every other provider credential — `AkahuItem`, `UpItem`, `RedbarkItem`, `LunchflowItem`, `PlaidItem`, `SimplefinItem` and so on. Making `MonobankItem` the one model that refuses to store a token when encryption is unconfigured would diverge from that app-wide contract and would break self-hosters who run without encryption keys today, for a reason that applies equally to every provider already shipped.

The case is also already surfaced to the user: `Settings::ProvidersController#set_encryption_warning_context` sets `@provider_setup_encryption_warning` when the deployment is self-hosted and `ActiveRecordEncryptionConfig.explicitly_configured?` is false, and the providers page renders that warning.

I do think the underlying point is fair — plaintext bank tokens at rest is a real risk. But tightening it belongs in `Encryptable`, for all 52 models at once, rather than in a single provider PR. Happy to open that as a separate issue if the maintainers want to pursue it.

File: app/models/monobank_item/importer.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (yaroslavshparuk) [reply app/models/monobank_item/importer.rb]: Fixed in 5fb96d58 — see the reply on the duplicate thread above for detail. Rate-limit results stay soft skips; hard failures now reach `import_transactions` and increment `failed`, and the forward window is persisted before the failure propagates.

File: app/models/monobank_item/syncer.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (yaroslavshparuk) [reply app/models/monobank_item/syncer.rb]: Fixed in 5fb96d58 — see the reply on the duplicate thread above. All five stages go through `I18n.t`, `count` is interpolated, and Ukrainian plural forms are supplied.

File: app/models/provider/monobank.rb [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+      begin
+        yield
+      rescue *RETRYABLE_ERRORS => e
+        retries += 1
+        if retries <= max_retries
+          delay = calculate_retry_delay(retries)
+          Rails.logger.warn(
+            "Monobank API: #{operation_name} failed (attempt #{retries}/#{max_retries}): " \
+            "#{e.class}: #{e.message}. Retrying in #{delay}s..."
+          )
```
[the remark this answers is not in the document]
Comment (yaroslavshparuk) [reply app/models/provider/monobank.rb:250]: `Provider::Monobank` is a bare HTTP client — `initialize(access_token)` and nothing else. It has no `family` and no `account_provider`, and those are precisely the fields that make a `DebugLogEntry` filterable for support, so entries written from here would land in `/settings/debug` unattributed.

None of the sibling clients log there either: `Provider::Redbark`, `Provider::Up` and `Provider::Akahu` use `Rails.logger` exclusively (0 `DebugLogEntry` calls between them). The convention is applied one layer up, where the context exists — and this PR already does that.

Concretely, every one of these four sites raises a typed `Provider::Monobank::Error`, and `MonobankItem::Importer#fetch_window` catches it and calls `capture_sync_error`, which emits a `DebugLogEntry` carrying `family`, `account_provider`, `provider_key: "monobank"` and the failure code. `MonobankItem` and `MonobankItem::Syncer` do the same for their own paths. So these failures do reach `/settings/debug` — recorded at the layer that knows whose connection they belong to.

The remaining one is the retry-attempt warning in `with_retries`, which is the transient local noise AGENTS.md explicitly says to leave in `Rails.logger`:

> Reserve raw Rails logging for low-value local noise; anything operators may need should go to the debug log.

File: app/views/monobank_items/setup_accounts.html.erb [hunk 10]
```diff
@@ -0,0 +1,73 @@
+<% content_for :title, t(".title") %>
+
+<%= render DS::Dialog.new do |dialog| %>
```
[the remark this answers is not in the document]
Comment (yaroslavshparuk) [reply app/views/monobank_items/setup_accounts.html.erb:3]: `DS::Dialog` already emits that frame itself, so this would nest a second one with a duplicate id.

In `app/components/DS/dialog.rb`: `variant` defaults to `"modal"`, `#frame` returns `@frame || variant`, and `#wrapper_element` renders `content_tag("turbo-frame", id: frame, &block)` unless the caller passes `disable_frame: true`. `setup_accounts.html.erb` calls `render DS::Dialog.new` with neither override, so the response is already wrapped in `<turbo-frame id="modal">` and the `frame: :modal` link in `_monobank_item.html.erb` resolves against it.

This also matches the sibling providers: `up_items`, `akahu_items` and `redbark_items` all render `DS::Dialog` unwrapped in their `setup_accounts.html.erb`.

(The `select_accounts` / `select_existing_account` views in this PR *do* carry an explicit `turbo_frame_tag "modal"` — that is copied from the akahu/lunchflow originals and is redundant for the same reason. Happy to strip it from all of them if you want the redundancy cleaned up.)

File: docs/hosting/monobank.md [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+and statements. You bring your own token, so Sure talks to monobank directly on your
+behalf — no third-party aggregator sits in between.
+
+> [!NOTE]
+> This integration uses monobank's **personal** API. Its documentation states that a
+> service which stores other people's data on its own servers must use the corporate
+> (service provider) API instead, and that programs used by clients personally — where
+> the data never passes through the developer's nodes — do not. A self-hosted Sure
+> instance is the latter. If you run Sure as a hosted service for other people, read
+> [monobank's API terms](https://api.monobank.ua/docs/index.html) first.
```
[the remark this answers is not in the document]
Comment (yaroslavshparuk) [reply docs/hosting/monobank.md:13]: Flagging this one for a maintainer decision rather than deciding it unilaterally, because it is a product/policy call about what the hosted instance offers.

The ToS reading is already documented at the top of this file:

> This integration uses monobank's **personal** API. Its documentation states that a service which stores other people's data on its own servers must use the corporate (service provider) API instead ... If you run Sure as a hosted service for other people, read monobank's API terms first.

What is not there is a hard code gate. No other per-family, bring-your-own-credential provider gates on `app_mode` — Akahu, Up, Redbark, Lunchflow and SimpleFIN are all available in managed mode — so adding one for monobank alone is a decision about your hosted offering rather than a consistency fix I should make on my own.

If you want the gate, I will add it: a `Rails.configuration.app_mode.self_hosted?` guard on the panel in `Settings::ProvidersController` plus a validation on `MonobankItem` so the token cannot be persisted in managed mode, and I will sharpen the doc wording to match. Just confirm that is the policy you want and I will push it.

File: docs/hosting/monobank.md [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+> The personal API is not available to clients under 16. Data for a child's accounts is
+> available from the parent's account.
+
+## 1. Get Your Personal Token
+
+1. Go to [api.monobank.ua](https://api.monobank.ua/).
+2. Sign in by scanning the QR code with the monobank app.
+3. Generate a personal token and copy it. **It is shown only once.**
+
+The token is read-only: it can list your accounts and read statements, nothing else.
```
[the remark this answers is not in the document]
Comment (yaroslavshparuk) [reply docs/hosting/monobank.md]: Fixed in 5fb96d58. You are right — `POST /personal/webhook` takes the same `X-Token`, so "read-only ... nothing else" was wrong. Reworded to say the token can list accounts, read statements and register a statement webhook, and that it cannot move money.

File: test/controllers/monobank_items_controller_test.rb [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+    @monobank_account = @monobank_item.monobank_accounts.create!(
+      name: "Black card ·1234",
+      account_id: "acc_1",
+      currency: "UAH",
+      account_kind: "card",
+      account_type: "black"
+    )
+  end
+
+  test "create saves the connection and starts a sync" do
```
[the remark this answers is not in the document]
Comment (yaroslavshparuk) [reply test/controllers/monobank_items_controller_test.rb:24]: Fixed in 5fb96d58.

- `create` expects `SyncJob.perform_later` once and additionally asserts the new connection has exactly one `Sync`.
- `sync` expects `SyncJob.perform_later` with a `Sync` whose `syncable` is the target connection.
- `destroy` expects `DestroyJob.perform_later` with the connection instead of stubbing it.

One correction to the suggestion: `sync_later` enqueues `SyncJob.perform_later(sync)` with the `Sync` record, not the connection, so the expectations match on `sync.syncable` rather than on the item itself.

File: test/encryption_verification_test.rb [hunk 14]
```diff
@@ -320,6 +320,48 @@ class EncryptionVerificationTest < ActiveSupport::TestCase
...
   end
 
+  test "monobank item credentials and payloads are encrypted" do
+    skip "No monobank items in fixtures" unless MonobankItem.any?
+
+    item = MonobankItem.first
+    original_payload = item.raw_payload
+
+    # Should be able to read
+    assert item.access_token.present? || item.raw_payload.present?
```
[the remark this answers is not in the document]
Comment (yaroslavshparuk) [reply test/encryption_verification_test.rb]: Agreed, and fixed in 5fb96d58 — the round trip proved nothing about what is on disk.

Both tests now write a known probe value, assert the accessor returns it, and then read the column straight through `ActiveRecord::Base.connection` and assert the stored bytes do not contain it. Added a small `assert_column_not_plaintext` helper for that, and extended the account test to cover `masked_pan` and `iban` as well as both payload columns.

**Comment by jjmata** [comment]
Are you using the Rails generator for this, @yaroslavshparuk?  Also, we're thinking of refactoring the `Provider` interface, swing by Discord where we are discussing it, in the #banks channel.

**Comment by yaroslavshparuk** [comment]
> Are you using the Rails generator for this, @yaroslavshparuk? Also, we're thinking of refactoring the `Provider` interface, swing by Discord where we are discussing it, in the #banks channel.

Thanks @jjmata 
Not sure if the agent ran the generator, but the layout seems matches provider:family --type=banking output. Or it is missing something?
Rest is Monobank specific, mostly their rate limit (one call a minute). 
Given the refactor talk, should I hold off or keep going?

File: app/models/provider/monobank.rb
Comment (jjmata) [inline app/models/provider/monobank.rb:1]: CodeRabbit already flagged the `Rails.logger` calls around the retry-warning branch (~L230-233) to convert to `DebugLogEntry.capture`, per the project's provider-diagnostics convention. The same gap exists a bit further down in this file and wasn't covered by that comment:

- `handle_response`'s `else` branch: `Rails.logger.error "Monobank API: Unexpected response status=#{response.code}"`
- `parse_response_body`'s rescue: `Rails.logger.error "Monobank API: Failed to parse response"`

Both are exactly the kind of "provider sync failure" diagnostic `DebugLogEntry.capture` exists for (per `CLAUDE.md`: "Prefer `DebugLogEntry.capture(...)` over `Rails.logger.*` for provider sync/import failures... Record support-relevant incidents in the super-admin `/settings/debug` UI"), and sibling files in this same PR (`monobank_item/importer.rb`, `monobank_item.rb`, the controller) do call `DebugLogEntry.capture` for equivalent errors. Worth sweeping this file for the remaining raw-logger spots while addressing the earlier comment, rather than fixing only the one line range.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by jjmata — COMMENTED** [review]
Reviewed against `CLAUDE.md` conventions (fat models/skinny controllers, `Current.family`/`Current.user`, i18n, Minitest+fixtures, minimal dependencies, provider integration shape) with a focus on the credential/API-key handling, error handling, and provider-abstraction-reuse angles called out for this new bank integration. Note that CodeRabbit and Codex already left extensive line-level feedback on this PR (encryption-guard readiness, hardcoded strings, sync-health wiring, view domain logic, rescue collapsing, managed-mode/ToS scoping, `RecordNotUnique` handling) — I did not re-flag those and instead checked areas not yet covered:

- **Dependencies**: no `Gemfile` changes; the provider uses `HTTParty` (already a dependency, same as other providers), consistent with "push Rails to its limits before adding a gem."
- **`Current.user`/`Current.family`**: controller and models use `Current.family` throughout; no `current_user`/`current_family` usage found.
- **Credential handling**: `access_token` is encrypted at rest via the same `encryption_ready?` guard pattern used by every other provider (Up, SimpleFIN, Plaid, etc.) — not a regression specific to this PR. No occurrence of `access_token` reaching a log line or `DebugLogEntry` metadata hash anywhere in the diff; `resolve_url` refuses to send the token to any host but `api.monobank.ua`.
- **Provider abstraction reuse**: `MonobankAccount::Transactions::CategoryMatcher`/`CategoryTaxonomy` explicitly mirror the existing Up/Plaid per-provider matchers rather than introducing a new pattern — consistent with (if not more DRY than) precedent, so not a duplication concern.
- **VCR**: no cassettes added, but this matches the existing convention for bring-your-own-token providers — `test/models/provider/simplefin_test.rb` also stubs the HTTP layer directly rather than recording cassettes (VCR is used for Plaid/Stripe/OpenAI-style flows instead).
- **One real gap found**: left an inline comment on `app/models/provider/monobank.rb` — a couple of `Rails.logger.*` calls in `handle_response`/`parse_response_body` weren't included in CodeRabbit's existing note about converting this file's retry logging to `DebugLogEntry.capture`.

Nothing else rose to the level of a blocking concern beyond what's already posted on the thread.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/models/provider/monobank.rb
Comment (yaroslavshparuk) [reply app/models/provider/monobank.rb:1]: Done — swept in 0bd03213, with follow-ups in 46daddb0.

All four `Rails.logger` calls in this file now go through `DebugLogEntry.capture`: the two you named plus the retry pair CodeRabbit's line range covered.

First, a correction. I pushed back on the CodeRabbit version of this on the grounds that a bare HTTP client has no `family` or `account_provider` to attach, and that `MonobankItem::Importer` already records these failures with that context one layer up. The second half holds; the conclusion did not. `Provider::Lunchflow`, `Provider::Questrade` and `Provider::Trading212` are bare clients that capture from exactly these paths, and AGENTS.md says to attach `family` *when available* rather than treating its absence as a reason not to log. I had compared only against Up, Akahu and Redbark — the smaller and less current sample.

What the client contributes that the importer genuinely cannot see: the HTTP status behind a `:fetch_failed`, the retry attempt and delay behind a `:network_error`, and the byte size of a body that would not parse. `#get` now threads the operation name down through `handle_response` into `parse_response_body`, so an entry names the endpoint that failed. The response body is never recorded — it carries statement PII.

**One deliberate exception to the sweep**, so it does not read as an oversight: the retry-attempt branch keeps its `Rails.logger.warn` alongside the debug entry. That mirrors `Provider::Lunchflow#with_retries`, which writes both, while its terminal-error path is debug-log only — and it matches AGENTS.md's "reserve raw Rails logging for low-value local noise". Without it this file was the only one in the monobank stack with no terminal trace of a failing request during `bin/dev`. The retry row also files under `provider_sync` rather than `provider_sync_error`, since a retry that then succeeds is not a failure and monobank already splits those two categories that way.

I had a second model review the change before pushing, which turned up a real gap next door: `MonobankItem::Importer#import_transactions` had a rescue writing only to `Rails.logger`. `fetch_window` captures provider errors itself, so what actually lands there is a *persistence* failure — an invalid snapshot, a cursor that will not save. It incremented the failure count and left nothing in `/settings/debug` to explain the resulting error badge. Fixed in 46daddb0 with a test.

Tests: an unexpected status writes an entry carrying status and operation; an unparseable body writes one carrying `body_bytes` but not the body; a storage failure in the importer reaches the debug log attributed to the right family and connection. Full suite 8410 runs, 0 failures; rubocop and brakeman clean.

**Comment by yaroslavshparuk** [comment]
Pushed four commits addressing the remaining review feedback.

**0bd03213 — provider client diagnostics to the debug log.** @jjmata's inline comment; replied in detail on that thread, including a correction to my earlier pushback on the CodeRabbit version of it.

**5a921038 — monobank sync health was never computed.** This one came out of CodeRabbit's "outside diff" note and is a real bug rather than a nit. `"monobank"` was in `PANEL_SYNCABLE_TYPES` but missing from `family_panel_items`, so `compute_provider_sync_health` skipped the key entirely: a monobank connection whose last sync failed or went stale rendered as `:ok` on `/settings/providers`, with no error badge and no "needs attention" grouping. One line, plus a test that a failed sync reaches `provider_sync_health` and lands the panel in `needs_attention`.

**55249e09 — collapsed the duplicate rescue** in `fetch_monobank_accounts_from_api`. The `Provider::Monobank::Error` and `StandardError` branches wrote the same `DebugLogEntry` and returned the same message, differing only in wording that `error_class` already carries.

**46daddb0 — gaps found by a second review pass.** `MonobankItem::Importer#import_transactions` had a rescue writing only to `Rails.logger`; it catches persistence failures rather than provider errors, so a failed snapshot or cursor save incremented a counter and left `/settings/debug` empty. Also replaced the `5a921038` test, which compared two hash literals and so stayed green for `"monobank" => nil`, with the behavioural one described above — verified failing both ways.

Two CodeRabbit nitpicks I have left alone deliberately: the `@monobank_sync_stats_map` fallback in `_monobank_item.html.erb` and the `Current.family` fallback in `_monobank_panel.html.erb` are byte-for-byte the shape Up and Redbark already use. Changing them for monobank alone makes it the odd one out; happy to convert all of them together in a follow-up if that is wanted.

Still open and waiting on a preference: the Codex thread about lifting the repeated account-choice row into a `DS::*` primitive. That shape exists in five other files across Akahu, Up and onchain wallets, so it is a seven-call-site design-system change. I would rather do it as a focused follow-up than bury it in this diff, but I will do it here if you would prefer that.

`bin/rails test` 8410 runs, 0 failures, 0 errors · `bin/rubocop` clean · `bin/brakeman` 0 warnings.

**Comment by yaroslavshparuk** [comment]
Pushed 0f0d3175, one more spot from a final sweep of the importer.

`MonobankItem::Importer#mark_requires_update!` swallowed a failed status update with only a `Rails.logger` line. A connection whose token was rejected would keep rendering as healthy with nothing in `/settings/debug`. It now goes through `capture_sync_error` like every other rescue in the file, with a test that fails against the old code.

`bin/rails test` 8411 runs, 0 failures, 0 errors.

**Comment by yaroslavshparuk** [comment]
Hi @jjmata
I think all the review feedback is addressed and the branch is synced with main. Is there anything else you need from me to get it merged? Happy to wait for the Provider refactor and rebase if you'd prefer that. Thanks!

File: app/models/monobank_account/transactions/category_taxonomy.rb
Comment (jjmata) [inline app/models/monobank_account/transactions/category_taxonomy.rb:1]: The MCC code to category mapping is something that should be abstracted and generalized (Enable Banking and Redbark capture it, buried in `transactions.extra` JSONB as provider-namespaced blobs, or did that change?

Should be something to revisit as we refactor the `Providers` code.

**Review by jjmata — APPROVED** [review]
